File: 数据结构/树的直径.cpp
/* author: Cysheper */
#include<bits/stdc++.h>
using namespace std;
#define all(x) x.begin(), x.end()
#define unq_all(x) x.erase(unique(all(x)), x.end())
#define inf 0x3f3f3f3f
#define eps 1e-19
typedef long long ll;
typedef unsigned long long ull;
typedef long double ld;
const int N = 1e5 + 5;
vector<int> e[N];
int ans = 0, poi = 0;
int vis[N];
void dfs(int x, int dep) {
    vis[x] = 1;
    for(auto v : e[x]) {
        if(!vis[v]) dfs(v, dep+1);
    }
    if(dep > ans) {
        ans = dep;
        poi = x;
    }
}
void solve() {
    int n; cin >> n;
    int a, b;
    for(int i=0;i<n-1;++i) {
        cin >> a >> b;
        e[a].push_back(b);
        e[b].push_back(a);
    }
    dfs(1, 1);
    memset(vis, 0, sizeof(vis));
    dfs(poi, 1);
    cout << ans-1;
}
int main() {
    ios::sync_with_stdio(false); 
    cin.tie(0);
    cout.tie(0);
    solve();
    return 0;
}

File: Python模板/树的直径.py
e = [[] for i in range(1000)]
vis = [0 for i in range(1000)]
ans, poi = 0, 0
def dfs(x, dep):
    global ans
    vis[x] = 1
    for v in e[x]:
        if vis[x] == 0: dfs(v, dep + 1)
    if dep > ans:
        ans = dep
        poi = x
        
def solve():
    n = int(input())
    a, b = map(int, input().split())
    for i in range(0, n-1):
        a, b = map(int, input().split())
        e[a].append(b)
        e[b].append(a)
    dfs(1, 1)
    vis.clear()
    dfs(poi, 1)
    print(ans - 1)

def main():
    solve()
    
if __name__ == "__main__":
    main()


    